### PR TITLE
feat: advanced barrel exports optimization.

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
### Description
1. Finished nested namespace ref barrel exports optimization
2. Change eager inclusion in `tree_shaking.rs` into **lazy** inclusion.

### CODE REVIEW

File: crates/rolldown/src/ast_scanner/impl_visit.rs
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/impl_visit.rs]: Could you move this method to `mod.ts`.

File: crates/rolldown/src/ast_scanner/impl_visit.rs
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/impl_visit.rs:51]: Do we need to call `walk::walk_member_expression` here?

Comment (IWANABETHATGUY) [reply crates/rolldown/src/ast_scanner/impl_visit.rs:51]: No, if you call `walk::walk_member_expresion`, then `identifier` visitor will be called, and another referenced symbol will be added to this module. Visitor of  `member_expr_reference` and `identifier_reference` should be mutually exclusive

Comment (IWANABETHATGUY) [reply crates/rolldown/src/ast_scanner/impl_visit.rs:51]: I could add `enter_node` and `leave_node` if we need.

File: crates/rolldown/src/stages/link_stage/mod.rs
Comment (IWANABETHATGUY) [inline crates/rolldown/src/stages/link_stage/mod.rs:286]: All these `into` invocation, convert `symbol_ref` into a  `SymbolOrMemberExprRef::Symbol`

File: crates/rolldown/tests/esbuild/import_star/other_file_export_self_as_namespace_unused_es6/artifacts.snap
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/import_star/other_file_export_self_as_namespace_unused_es6/artifacts.snap:19]: esbuild: 
```js
================================================================================
TestOtherFileExportSelfAsNamespaceUnusedES6
---------- /out.js ----------
// foo.js
var foo = 123;
export {
  foo
};

```

File: crates/rolldown/tests/esbuild/import_star/import_star_export_star_as_unused/artifacts.snap
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/import_star/import_star_export_star_as_unused/artifacts.snap:12]:
```js
================================================================================
TestImportStarExportStarAsUnused
---------- /out.js ----------
// entry.js
var foo = 234;
console.log(foo);

```

File: crates/rolldown/src/stages/link_stage/tree_shaking.rs
Comment (IWANABETHATGUY) [inline crates/rolldown/src/stages/link_stage/tree_shaking.rs:115]: Resolve top level member expr chain reference.

File: crates/rolldown/tests/esbuild/dce/import_re_export_of_namespace_import/artifacts.snap
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/dce/import_re_export_of_namespace_import/artifacts.snap:28]: Only `foo` is used.

File: crates/rolldown/tests/esbuild/dce/tree_shaking_import_identifier/artifacts.snap
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/dce/tree_shaking_import_identifier/artifacts.snap:26]: Only keep is used.
here is the result of webpack: 
```js
/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
var __webpack_exports__ = {};

;// CONCATENATED MODULE: ./a/b.mjs
class Base {}

;// CONCATENATED MODULE: ./a/a.mjs

class Keep extends Base {}
class REMOVE extends Base {}


;// CONCATENATED MODULE: ./a/index.js

new Keep();

/******/ })()
;

```

Comment (IWANABETHATGUY) [reply crates/rolldown/tests/esbuild/dce/tree_shaking_import_identifier/artifacts.snap:26]: the namespace object ref will be eliminated after we merge https://github.com/rolldown/rolldown/pull/1381

File: crates/rolldown/Cargo.toml
Comment (hyf0) [inline crates/rolldown/Cargo.toml]: Please use workspace so we could unify the version across crates.

File: crates/rolldown/src/ast_scanner/impl_visit.rs
Comment (hyf0) [reply crates/rolldown/src/ast_scanner/impl_visit.rs:51]: Ok. I prefer not to use `enter_node` and `leave_node` mechanism in any situation. It's hard to control and maintain it's flow.

Comment (hyf0) [reply crates/rolldown/src/ast_scanner/impl_visit.rs:51]: I just look it again. Could you rewrite it like

```rust
fn visit_member_expression{
  if MemberExpression::StaticMemberExpression ... {
    // ...
    // Bailout for not creating another referenced symbol
    return
  }

  // nothing special, just keep walking
  walk::visit_member_expression(self)
}
```

The current logic is complex to see when we should keep walking when we should bailout. The other visitor's function is writing like this way. It would be good if we could keep this style.

- https://github.com/rolldown/rolldown/blob/aef45e09fa93e7df8f18422fb8d5c47fae0c77f0/crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs#L356-L360
- https://github.com/rolldown/rolldown/blob/aef45e09fa93e7df8f18422fb8d5c47fae0c77f0/crates/rolldown/src/module_finalizers/scope_hoisting/impl_visit_mut.rs#L421-L464
- https://github.com/rolldown/rolldown/blob/aef45e09fa93e7df8f18422fb8d5c47fae0c77f0/crates/rolldown/src/ast_scanner/impl_visit.rs#L74-L79

File: crates/rolldown_common/src/types/symbol_ref.rs
Comment (hyf0) [inline crates/rolldown_common/src/types/symbol_ref.rs]: Is this meaningful?

File: crates/rolldown_common/src/types/symbol_ref.rs
Comment (hyf0) [inline crates/rolldown_common/src/types/symbol_ref.rs]:
```suggestion
  pub fn symbol_ref(&self) -> &SymbolRef {
```

I think it's ok you just call it `symbol` or `symbol_ref`. Because when we are referring `symbol` in `test.a.b`, the symbol would still be `test`. `.a` and `.b` might point to some symbols, but for the expression `test.a.b` itself, the symbol would only be `test`.

File: crates/rolldown_common/src/types/symbol_ref.rs
Comment (hyf0) [inline crates/rolldown_common/src/types/symbol_ref.rs:39]: Discussion only: Do you think if it would be better we just call it `SymbolRef`?

```rust
pub struct SymbolRef {
  owner: ...,
  symbol: ...,
  chains: ..., or static_members:...
}

```

Is there any case that we will only benefit from using enum?

**Comment by hyf0** [comment]
Nice work. I have comment suggested changes and other parts are all fine. We are good to merge once we solve above comments.

https://github.com/rolldown/rolldown/pull/1374#discussion_r1641907286 is only an discussion and not required to be solved in this PR.

File: crates/rolldown_common/src/types/symbol_ref.rs
Comment (IWANABETHATGUY) [reply crates/rolldown_common/src/types/symbol_ref.rs:39]: There is already a struct called `SymbolRef`, do you mean change the original `SymbolRef` sturct? 
the downsides are: 
1. the memory usage will from **8** bytes to **24**
2. the `SymbolRef` will not `Copy` any more

Comment (IWANABETHATGUY) [reply crates/rolldown_common/src/types/symbol_ref.rs:39]: Since we used  `SymbolRef` a lot, so the performance difference should be considerable.

Comment (hyf0) [reply crates/rolldown_common/src/types/symbol_ref.rs:39]: >  do you mean change the original SymbolRef sturct?

Yeah.

> the memory usage will from 8 bytes to 24
> the SymbolRef will not Copy any more

I'm only think it in aspect of if it make code cleaner. Feel like "Not copy" will make the dx worse. Ok. Let's keep it how it is now and see how it envolves.

**Review by hyf0 — APPROVED** [review]

**Pull request merged**
